File: Voter Deduplication System/preprocessing/normalization.py
import re

def normalize_name(name):
    if not name:
        return ""
    # Convert to uppercase, remove non-alphanumeric (except space), strip whitespace
    name = str(name).upper()
    name = re.sub(r'[^A-Z0-9\s]', '', name)
    name = " ".join(name.split())
    return name

def normalize_dob(dob):
    if not dob:
        return ""
    # Keep only digits and hyphens/slashes, then standardize to YYYY-MM-DD if possible
    dob = str(dob).strip()
    return dob

def normalize_mobile(mobile):
    if not mobile:
        return ""
    # Keep only digits, remove leading 0 or +91
    mobile = re.sub(r'\D', '', str(mobile))
    if len(mobile) > 10:
        if mobile.startswith("91"):
            mobile = mobile[2:]
        elif mobile.startswith("0"):
            mobile = mobile[1:]
    return mobile[-10:] if len(mobile) >= 10 else mobile

def normalize_address(address):
    if not address:
        return ""
    # Convert to uppercase, remove punctuation, strip whitespace
    address = str(address).upper()
    address = re.sub(r'[^A-Z0-9\s]', ' ', address)
    address = " ".join(address.split())
    return address

File: backend/utils/normalization.js
export const normalizeName = (name) => {
    if (!name) return "";
    return name.toString().toUpperCase()
        .replace(/[^A-Z0-9\s]/g, '')
        .trim()
        .replace(/\s+/g, ' ');
};

export const normalizeDob = (dob) => {
    if (!dob) return "";
    // If it's a Date object, format to YYYY-MM-DD
    if (dob instanceof Date) {
        return dob.toISOString().split('T')[0];
    }
    return dob.toString().trim();
};

export const normalizeMobile = (mobile) => {
    if (!mobile) return "";
    let clean = mobile.toString().replace(/\D/g, '');
    if (clean.length > 10) {
        if (clean.startsWith('91')) {
            clean = clean.slice(2);
        } else if (clean.startsWith('0')) {
            clean = clean.slice(1);
        }
    }
    return clean.length >= 10 ? clean.slice(-10) : clean;
};

export const normalizeAddress = (address) => {
    if (!address) return "";
    return address.toString().toUpperCase()
        .replace(/[^A-Z0-9\s]/g, ' ')
        .trim()
        .replace(/\s+/g, ' ');
};
